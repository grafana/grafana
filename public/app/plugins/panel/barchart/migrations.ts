import { FieldMatcherID, PanelTypeChangedHandler, ReducerID } from '@grafana/data';
import { AxisPlacement } from '@grafana/ui';

/*
 * This is called when the panel changes from another panel
 */
export const changeToBarChartPanelMigrationHandler: PanelTypeChangedHandler = (panel, prevPluginId, prevOptions) => {
  if (prevPluginId === 'graph') {
    const graphOptions: GraphOptions = prevOptions.angular;

    const fieldConfig = panel.fieldConfig ?? { defaults: {}, overrides: [] };

    if (graphOptions.xaxis?.mode === 'series') {
      const transformations = panel.transformations || [];
      transformations.push(
        {
          id: 'reduce',
          options: {
            reducers: getReducer(graphOptions.xaxis?.values),
          },
        },
        {
          id: 'transpose',
          options: {},
        }
      );

      panel.transformations = transformations;

      // temporary, until we have a bar chart with per bar labels
      fieldConfig.overrides.push({
        matcher: {
          id: FieldMatcherID.byName,
          options: 'Field',
        },
        properties: [
          {
            id: 'custom.axisPlacement',
            value: AxisPlacement.Hidden,
          },
        ],
      });

      panel.fieldConfig = fieldConfig;
      panel.options = {
        ...panel.options,
        groupWidth: 1,
      };
    }
  }

  return {};
};

// same as grafana-ui/src/components/SingleStatShared/SingleStatBaseOptions.ts
const getReducer = (reducers: string[] | undefined) => {
  const transformReducers: string[] = [];

  reducers?.forEach((reducer) => {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    if (!Object.values(ReducerID).includes(reducer as ReducerID)) {
      if (reducer === 'current') {
        transformReducers.push(ReducerID.lastNotNull);
      } else if (reducer === 'total') {
        transformReducers.push(ReducerID.sum);
      } else if (reducer === 'avg') {
        transformReducers.push(ReducerID.mean);
      }
    } else {
      transformReducers.push(reducer);
    }
  });

  return reducers ? transformReducers : [ReducerID.sum];
};

interface GraphOptions {
  xaxis: {
    mode: 'series' | 'time' | 'histogram';
    values?: string[];
  };
}
