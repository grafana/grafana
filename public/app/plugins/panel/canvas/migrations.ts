import { PanelModel, OneClickMode } from '@grafana/data';

import { Options } from './panelcfg.gen';

export const canvasMigrationHandler = (panel: PanelModel): Partial<Options> => {
  const pluginVersion = panel?.pluginVersion ?? '';

  // Rename text-box to rectangle
  // Initial plugin version is empty string for first migration
  if (pluginVersion === '') {
    const root = panel.options?.root;
    if (root?.elements) {
      for (const element of root.elements) {
        if (element.type === 'text-box') {
          element.type = 'rectangle';
        }
      }
    }
  }

  if (pluginVersion.startsWith('11.0')) {
    // Migration for v11.0 for ellipse element refactor: https://github.com/grafana/grafana/pull/84205
    const root = panel.options?.root;
    if (root?.elements) {
      for (const element of root.elements) {
        if (element.type === 'ellipse') {
          // Take existing ellipse specific background and border config and apply it to the element's general background and border config
          if (element.config.backgroundColor) {
            element.background = element.config.backgroundColor;
            delete element.config.backgroundColor;
          }
          if (element.config.borderColor) {
            element.border.color = element.config.borderColor;
            delete element.config.borderColor;
          }
          if (element.config.width) {
            element.border.width = element.config.width;
            delete element.config.width;
          }
        }
      }
    }
  }

  if (parseFloat(pluginVersion) <= 11.3) {
    const root = panel.options?.root;
    if (root?.elements) {
      for (const element of root.elements) {
        // migrate oneClickLinks to oneClickMode
        if (element.oneClickLinks) {
          element.oneClickMode = OneClickMode.Link;
          delete element.oneClickLinks;
        }

        // migrate action options to new format (fetch)
        if (element.actions) {
          for (const action of element.actions) {
            if (action.options) {
              action.fetch = { ...action.options };
              delete action.options;
            }
          }
        }
      }
    }
  }

  return panel.options;
};
