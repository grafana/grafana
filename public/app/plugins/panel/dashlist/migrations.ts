import { PanelModel } from '@grafana/data';
import { getBackendSrv } from '@grafana/runtime';
import { FolderDTO } from 'app/types';

import { Options } from './panelcfg.gen';

async function getFolderUID(folderID: number): Promise<string> {
  // folderID 0 is always the fake General/Dashboards folder, which always has a UID of empty string
  if (folderID === 0) {
    return '';
  }

  const folderDTO = await getBackendSrv().get<FolderDTO>(`/api/folders/id/${folderID}`, undefined, undefined, {
    showErrorAlert: false,
  });

  return folderDTO.uid;
}

export interface AngularModel {
  /** @deprecated */
  starred?: boolean;
  /** @deprecated */
  recent?: boolean;
  /** @deprecated */
  search?: boolean;
  /** @deprecated */
  headings?: boolean;
  /** @deprecated */
  limit?: number;
  /** @deprecated */
  query?: string;
  /** @deprecated */
  folderId?: number;
  /** @deprecated */
  tags?: string[];
}

export async function dashlistMigrationHandler(panel: PanelModel<Options> & AngularModel) {
  // Convert old angular model to new react model
  const newOptions: Options = {
    ...panel.options,
    showStarred: panel.options.showStarred ?? panel.starred,
    showRecentlyViewed: panel.options.showRecentlyViewed ?? panel.recent,
    showSearch: panel.options.showSearch ?? panel.search,
    showHeadings: panel.options.showHeadings ?? panel.headings,
    maxItems: panel.options.maxItems ?? panel.limit,
    query: panel.options.query ?? panel.query,
    folderId: panel.options.folderId ?? panel.folderId,
    tags: panel.options.tags ?? panel.tags,
  };

  // Delete old angular properties
  const previousVersion = parseFloat(panel.pluginVersion || '6.1');
  if (previousVersion < 6.3) {
    const oldProps = ['starred', 'recent', 'search', 'headings', 'limit', 'query', 'folderId'] as const;
    oldProps.forEach((prop) => delete panel[prop]);
  }

  // Convert the folderId to folderUID. Uses the API to do the conversion.
  if (newOptions.folderId !== undefined) {
    const folderId = newOptions.folderId;

    // If converting ID to UID fails, the panel will not be migrated and will show incorrectly
    try {
      const folderUID = await getFolderUID(folderId);
      newOptions.folderUID = folderUID;
      delete newOptions.folderId;
    } catch (err) {
      console.warn('Dashlist: Error migrating folder ID to UID', err);
    }
  }

  return newOptions;
}
