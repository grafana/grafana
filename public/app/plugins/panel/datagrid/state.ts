import {
  CellClickedEventArgs,
  GridColumnIcon,
  GridSelection,
  HeaderClickedEventArgs,
  Item,
  Rectangle,
  SizedGridColumn,
} from '@glideapps/glide-data-grid';

import { DataFrame, Field, FieldType, getFieldDisplayName } from '@grafana/data';

import { isDatagridEnabled } from './featureFlagUtils';
import {
  DatagridContextMenuData,
  DEFAULT_CONTEXT_MENU,
  DEFAULT_RENAME_INPUT_DATA,
  EMPTY_GRID_SELECTION,
  getCellWidth,
  RenameColumnInputData,
} from './utils';

interface DatagridState {
  columns: SizedGridColumn[];
  contextMenuData: DatagridContextMenuData;
  renameColumnInputData: RenameColumnInputData;
  gridSelection: GridSelection;
  columnFreezeIndex: number;
  toggleSearch: boolean;
  isResizeInProgress: boolean;
}

interface UpdateColumnsAction {
  type: DatagridActionType.updateColumns;
  payload: {
    frame: DataFrame;
  };
}

interface ColumnResizeStartAction {
  type: DatagridActionType.columnResizeStart;
  payload: {
    columnIndex: number;
    width: number;
  };
}

interface OpenCellContextMenuAction {
  type: DatagridActionType.openCellContextMenu;
  payload: {
    event: CellClickedEventArgs;
    cell: Item;
  };
}

interface OpenHeaderContextMenuAction {
  type: DatagridActionType.openHeaderContextMenu;
  payload: {
    event: HeaderClickedEventArgs;
    columnIndex: number;
  };
}

interface OpenHeaderDropdownMenuAction {
  type: DatagridActionType.openHeaderDropdownMenu;
  payload: {
    screenPosition: Rectangle;
    columnIndex: number;
    value: string;
  };
}

interface ColumnMoveAction {
  type: DatagridActionType.columnMove;
  payload: {
    from: number;
    to: number;
  };
}

interface MultipleCellsSelectedAction {
  type: DatagridActionType.multipleCellsSelected;
  payload: {
    selection: GridSelection;
  };
}

interface ColumnFreezeChangedAction {
  type: DatagridActionType.columnFreezeChanged;
  payload: {
    columnIndex: number;
  };
}

interface ColumnResizeEndAction {
  type: DatagridActionType.columnResizeEnd;
}
interface ShowColumnRenameInputAction {
  type: DatagridActionType.showColumnRenameInput;
}
interface HideColumnRenameInputAction {
  type: DatagridActionType.hideColumnRenameInput;
}
interface CloseContextMenuAction {
  type: DatagridActionType.closeContextMenu;
}
interface CloseSearchAction {
  type: DatagridActionType.closeSearch;
}
interface OpenSearchAction {
  type: DatagridActionType.openSearch;
}
interface GridSelectionClearedAction {
  type: DatagridActionType.gridSelectionCleared;
}
interface ColumnFreezeResetAction {
  type: DatagridActionType.columnFreezeReset;
}

export type DatagridAction =
  | UpdateColumnsAction
  | ColumnResizeStartAction
  | OpenCellContextMenuAction
  | OpenHeaderContextMenuAction
  | OpenHeaderDropdownMenuAction
  | ColumnMoveAction
  | MultipleCellsSelectedAction
  | ColumnFreezeChangedAction
  | ColumnResizeEndAction
  | ShowColumnRenameInputAction
  | HideColumnRenameInputAction
  | CloseContextMenuAction
  | CloseSearchAction
  | OpenSearchAction
  | GridSelectionClearedAction
  | ColumnFreezeResetAction;

export enum DatagridActionType {
  columnResizeStart = 'columnResizeStart',
  columnResizeEnd = 'columnResizeEnd',
  columnMove = 'columnMove',
  updateColumns = 'updateColumns',
  showColumnRenameInput = 'showColumnRenameInput',
  hideColumnRenameInput = 'hideColumnRenameInput',
  openCellContextMenu = 'openCellContextMenu',
  openHeaderContextMenu = 'openHeaderContextMenu',
  openHeaderDropdownMenu = 'openHeaderDropdownMenu',
  closeContextMenu = 'closeContextMenu',
  multipleCellsSelected = 'multipleCellsSelected',
  gridSelectionCleared = 'gridSelectionCleared',
  columnFreezeReset = 'columnFreezeReset',
  columnFreezeChanged = 'columnFreezeChanged',
  openSearch = 'openSearch',
  closeSearch = 'closeSearch',
}

export const initialState: DatagridState = {
  columns: [],
  contextMenuData: DEFAULT_CONTEXT_MENU,
  renameColumnInputData: DEFAULT_RENAME_INPUT_DATA,
  gridSelection: EMPTY_GRID_SELECTION,
  columnFreezeIndex: 0,
  toggleSearch: false,
  isResizeInProgress: false,
};

const typeToIconMap: Map<string, GridColumnIcon> = new Map([
  [FieldType.number, GridColumnIcon.HeaderNumber],
  [FieldType.string, GridColumnIcon.HeaderTextTemplate],
  [FieldType.boolean, GridColumnIcon.HeaderBoolean],
  [FieldType.time, GridColumnIcon.HeaderDate],
  [FieldType.other, GridColumnIcon.HeaderReference],
]);

export const datagridReducer = (state: DatagridState, action: DatagridAction): DatagridState => {
  let columns: SizedGridColumn[] = [];

  switch (action.type) {
    case DatagridActionType.columnResizeStart:
      columns = [...state.columns];
      const columnResizeStartPayload = action.payload;

      columns[columnResizeStartPayload.columnIndex] = {
        ...state.columns[columnResizeStartPayload.columnIndex],
        width: columnResizeStartPayload.width,
      };

      return {
        ...state,
        columns,
        isResizeInProgress: true,
      };
    case DatagridActionType.columnMove:
      columns = [...state.columns];
      const columnMovePayload = action.payload;

      const widthFrom = state.columns[columnMovePayload.from].width;

      let fromColumn = columns.splice(columnMovePayload.from, 1)[0];

      fromColumn = {
        ...fromColumn,
        width: widthFrom,
      };

      columns.splice(columnMovePayload.to, 0, fromColumn);

      return {
        ...state,
        columns,
      };
    case DatagridActionType.columnResizeEnd:
      return {
        ...state,
        isResizeInProgress: false,
      };
    case DatagridActionType.updateColumns:
      const updateColumnsPayload = action.payload;

      columns = [
        ...updateColumnsPayload.frame.fields.map((field: Field, index: number) => {
          // find column by field name and update width in new set. We cannot use index because
          // if a column gets deleted we don't know the correct index anymore
          const width = state.columns.find((column) => column.title === field.name)?.width;
          const displayName = getFieldDisplayName(field, updateColumnsPayload.frame);

          return {
            title: displayName,
            width: width ?? getCellWidth(field),
            icon: typeToIconMap.get(field.type),
            hasMenu: isDatagridEnabled(),
            trailingRowOptions: { targetColumn: --index },
          };
        }),
      ];

      return {
        ...state,
        columns,
      };
    case DatagridActionType.showColumnRenameInput:
      return {
        ...state,
        renameColumnInputData: {
          ...state.renameColumnInputData,
          isInputOpen: true,
        },
      };
    case DatagridActionType.hideColumnRenameInput:
      return {
        ...state,
        renameColumnInputData: {
          ...state.renameColumnInputData,
          isInputOpen: false,
        },
      };
    case DatagridActionType.openCellContextMenu:
      const openCellContextMenuPayload = action.payload;
      const cellEvent: CellClickedEventArgs = openCellContextMenuPayload.event;
      const cell: Item = openCellContextMenuPayload.cell;

      return {
        ...state,
        contextMenuData: {
          x: cellEvent.bounds.x + cellEvent.localEventX,
          y: cellEvent.bounds.y + cellEvent.localEventY,
          column: cell[0] === -1 ? undefined : cell[0], //row numbers,
          row: cell[1],
          isContextMenuOpen: true,
          isHeaderMenu: false,
        },
      };
    case DatagridActionType.openHeaderContextMenu:
      const openHeaderContextMenuPayload = action.payload;
      const headerEvent: HeaderClickedEventArgs = openHeaderContextMenuPayload.event;

      return {
        ...state,
        contextMenuData: {
          x: headerEvent.bounds.x + headerEvent.localEventX,
          y: headerEvent.bounds.y + headerEvent.localEventY,
          column: openHeaderContextMenuPayload.columnIndex,
          row: undefined, //header
          isContextMenuOpen: true,
          isHeaderMenu: false,
        },
      };
    case DatagridActionType.openHeaderDropdownMenu:
      const openHeaderDropdownMenuPayload = action.payload;
      const screenPosition: Rectangle = openHeaderDropdownMenuPayload.screenPosition;

      return {
        ...state,
        contextMenuData: {
          x: screenPosition.x + screenPosition.width,
          y: screenPosition.y + screenPosition.height,
          column: openHeaderDropdownMenuPayload.columnIndex,
          row: undefined, //header
          isContextMenuOpen: true,
          isHeaderMenu: true,
        },
        renameColumnInputData: {
          x: screenPosition.x,
          y: screenPosition.y,
          width: screenPosition.width,
          height: screenPosition.height,
          columnIdx: openHeaderDropdownMenuPayload.columnIndex,
          isInputOpen: false,
          inputValue: openHeaderDropdownMenuPayload.value,
        },
      };
    case DatagridActionType.closeContextMenu:
      return {
        ...state,
        contextMenuData: {
          isContextMenuOpen: false,
        },
      };
    case DatagridActionType.closeSearch:
      return {
        ...state,
        toggleSearch: false,
      };
    case DatagridActionType.openSearch:
      return {
        ...state,
        toggleSearch: true,
      };
    case DatagridActionType.multipleCellsSelected:
      const multipleCellsSelectedPayload = action.payload;

      return {
        ...state,
        gridSelection: multipleCellsSelectedPayload.selection,
      };
    case DatagridActionType.gridSelectionCleared:
      return {
        ...state,
        gridSelection: EMPTY_GRID_SELECTION,
      };
    case DatagridActionType.columnFreezeReset:
      return {
        ...state,
        columnFreezeIndex: 0,
      };
    case DatagridActionType.columnFreezeChanged:
      const columnFreezeChangedPayload = action.payload;

      return {
        ...state,
        columnFreezeIndex: columnFreezeChangedPayload.columnIndex,
      };
    default:
      return state;
  }
};
