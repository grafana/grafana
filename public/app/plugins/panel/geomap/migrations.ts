import { cloneDeep } from 'lodash';

import {
  FieldConfigSource,
  PanelModel,
  PanelTypeChangedHandler,
  Threshold,
  ThresholdsMode,
  fieldReducers,
  FrameGeometrySourceMode,
  DataTransformerConfig,
  DataTransformerID,
} from '@grafana/data';
import { ResourceDimensionMode } from '@grafana/schema';

import { defaultMarkersConfig, MarkersConfig } from './layers/data/markersLayer';
import { getMarkerAsPath } from './style/markers';
import { defaultStyleConfig } from './style/types';
import { Options, TooltipMode } from './types';
import { MapCenterID } from './view';

/**
 * This is called when the panel changes from another panel
 */
export const mapPanelChangedHandler: PanelTypeChangedHandler = (panel, prevPluginId, prevOptions, prevFieldConfig) => {
  // Changing from angular/worldmap panel to react/openlayers
  if (prevPluginId === 'grafana-worldmap-panel' && prevOptions.angular) {
    const { fieldConfig, options, xform } = worldmapToGeomapOptions({
      ...prevOptions.angular,
      fieldConfig: prevFieldConfig,
    });
    if (xform?.id?.length) {
      panel.transformations = panel.transformations ? [...panel.transformations, xform] : [xform];
    }
    panel.fieldConfig = fieldConfig; // Mutates the incoming panel
    return options;
  }

  return {};
};

export function worldmapToGeomapOptions(angular: any): {
  fieldConfig: FieldConfigSource;
  options: Options;
  xform?: DataTransformerConfig;
} {
  const fieldConfig: FieldConfigSource = {
    defaults: {},
    overrides: [],
  };

  const markersLayer = cloneDeep(defaultMarkersConfig);
  const options: Options = {
    view: {
      id: MapCenterID.Zero,
    },
    controls: {
      showZoom: true,
      mouseWheelZoom: Boolean(angular.mouseWheelZoom),
    },
    basemap: {
      type: 'default', // was carto
      name: 'Basemap',
    },
    layers: [markersLayer],
    tooltip: { mode: TooltipMode.Details },
  };

  let v = asNumber(angular.decimals);
  if (v) {
    fieldConfig.defaults.decimals = v;
  }

  // Set the markers range
  const style = markersLayer.config!.style;
  v = asNumber(angular.circleMaxSize);
  if (v) {
    style.size!.max = v;
  }
  v = asNumber(angular.circleMinSize);
  if (v) {
    style.size!.min = v;
  }

  let xform: DataTransformerConfig | undefined = undefined;
  const reducer = fieldReducers.getIfExists(angular.valueName);
  if (reducer && angular.locationData?.length) {
    xform = {
      id: DataTransformerID.reduce,
      options: {
        reducers: [reducer.id],
      },
    };

    switch (angular.locationData) {
      case 'countries':
      case 'countries_3letter':
        markersLayer.location = {
          mode: FrameGeometrySourceMode.Lookup,
          gazetteer: 'public/gazetteer/countries.json',
          lookup: undefined, // will default to first string field from reducer
        };
        break;

      case 'states':
        markersLayer.location = {
          mode: FrameGeometrySourceMode.Lookup,
          gazetteer: 'public/gazetteer/usa-states.json',
          lookup: undefined, // will default to first string field from reducer
        };
        break;
    }
  }

  // Convert thresholds and color values
  if (angular.thresholds && angular.colors) {
    const levels = angular.thresholds.split(',').map((strVale: string) => {
      return Number(strVale.trim());
    });

    // One more color than threshold
    const thresholds: Threshold[] = [];
    for (const color of angular.colors) {
      const idx = thresholds.length - 1;
      if (idx >= 0) {
        thresholds.push({ value: levels[idx], color });
      } else {
        thresholds.push({ value: -Infinity, color });
      }
    }

    fieldConfig.defaults.thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: thresholds,
    };
  }

  v = asNumber(angular.initialZoom);
  if (v) {
    options.view.zoom = v;
  }

  // mapCenter: 'Europe',
  // mapCenterLatitude: 46,
  // mapCenterLongitude: 14,
  //
  // Map center (from worldmap)
  const mapCenters: any = {
    '(0°, 0°)': MapCenterID.Zero,
    'North America': 'north-america',
    Europe: 'europe',
    'West Asia': 'west-asia',
    'SE Asia': 'se-asia',
    'Last GeoHash': MapCenterID.Coordinates, // MapCenterID.LastPoint,
  };
  options.view.id = mapCenters[angular.mapCenter];
  options.view.lat = asNumber(angular.mapCenterLatitude);
  options.view.lon = asNumber(angular.mapCenterLongitude);
  return { fieldConfig, options, xform };
}

function asNumber(v: unknown): number | undefined {
  const num = Number(v);
  return isNaN(num) ? undefined : num;
}

export const mapMigrationHandler = (panel: PanelModel): Partial<Options> => {
  const pluginVersion = panel?.pluginVersion ?? '';

  // before 8.3, only one layer was supported!
  if (pluginVersion.startsWith('8.1') || pluginVersion.startsWith('8.2')) {
    const layers = panel.options?.layers;
    if (layers?.length === 1) {
      const layer = panel.options.layers[0];
      if (layer?.type === 'markers' && layer.config) {
        // Moving style to child object
        const oldConfig = layer.config;
        const config: MarkersConfig = {
          style: cloneDeep(defaultStyleConfig),
          showLegend: Boolean(oldConfig.showLegend),
        };

        if (oldConfig.size) {
          config.style.size = oldConfig.size;
        }
        if (oldConfig.color) {
          config.style.color = oldConfig.color;
        }
        if (oldConfig.fillOpacity) {
          config.style.opacity = oldConfig.fillOpacity;
        }
        const symbol = getMarkerAsPath(oldConfig.shape);
        if (symbol) {
          config.style.symbol = {
            fixed: symbol,
            mode: ResourceDimensionMode.Fixed,
          };
        }
        return { ...panel.options, layers: [{ ...layer, config }] };
      }
    }
  }
  return panel.options;
};
