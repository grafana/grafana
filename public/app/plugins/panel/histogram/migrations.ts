import { isNil, omitBy } from 'lodash';

import { FieldConfigSource, PanelTypeChangedHandler } from '@grafana/data';
import {
  LegendDisplayMode,
  SortOrder,
  StackingMode,
  TooltipDisplayMode,
} from '@grafana/schema/dist/esm/common/common.gen';

import { defaultHistogramConfig } from './config';
import { FieldConfig as HistogramFieldConfig, Options } from './panelcfg.gen';

/*
 * This is called when the panel changes from another panel
 */
export const changeToHistogramPanelMigrationHandler: PanelTypeChangedHandler = (
  panel,
  prevPluginId,
  prevOptions,
  prevFieldConfig
) => {
  if (prevPluginId === 'graph') {
    const graphOptions: GraphOptions = prevOptions.angular;

    if (graphOptions.xaxis?.mode === 'histogram') {
      const { fieldConfig, options } = graphToHistogramOptions({
        ...prevOptions.angular,
        fieldConfig: prevFieldConfig,
      });

      panel.fieldConfig = fieldConfig; // Mutates the incoming panel

      return options;
    }
  }

  return {};
};

function graphToHistogramOptions(graphOptions: GraphOptions): {
  fieldConfig: FieldConfigSource;
  options: Options;
} {
  let histogramFieldConfig: HistogramFieldConfig = {};
  const options: Options = {
    legend: {
      displayMode: LegendDisplayMode.List,
      showLegend: true,
      placement: 'bottom',
      calcs: [],
    },
    tooltip: {
      mode: TooltipDisplayMode.Single,
      sort: SortOrder.None,
    },
    combine: false,
  };

  if (graphOptions.stack) {
    histogramFieldConfig.stacking = {
      mode: graphOptions.percentage ? StackingMode.Percent : StackingMode.Normal,
      group: defaultHistogramConfig.stacking!.group,
    };

    options.combine = false;
  }

  return {
    fieldConfig: {
      defaults: omitBy(
        {
          custom: histogramFieldConfig,
        },
        isNil
      ),
      overrides: [],
    },
    options,
  };
}

interface GraphOptions {
  stack?: boolean;
  percentage?: boolean;
  xaxis: {
    mode: 'series' | 'time' | 'histogram';
    values?: string[];
  };
}
