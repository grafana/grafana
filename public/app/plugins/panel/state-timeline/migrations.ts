import { isArray } from 'lodash';

import { FieldConfigSource, MappingType, PanelModel, ValueMap } from '@grafana/data';

import { FieldConfig, Options } from './panelcfg.gen';

// This is called when the panel changes from another panel
export const timelinePanelChangedHandler = (
  panel: PanelModel<Partial<Options>> | any,
  prevPluginId: string,
  prevOptions: any
) => {
  let options: Options = panel.options ?? {};

  // Changing from angular singlestat
  if (prevPluginId === 'natel-discrete-panel' && prevOptions.angular) {
    const oldOptions = prevOptions.angular;
    const fieldConfig: FieldConfigSource = panel.fieldConfig ?? { defaults: {}, overrides: [] };

    if (oldOptions.units) {
      fieldConfig.defaults.unit = oldOptions.units;
    }

    const custom: FieldConfig = {
      fillOpacity: 100,
      lineWidth: 0,
    };
    fieldConfig.defaults.custom = custom;
    options.mergeValues = true;

    // Convert mappings
    const valuemap: ValueMap = { type: MappingType.ValueToText, options: {} };
    fieldConfig.defaults.mappings = [valuemap];

    if (isArray(oldOptions.colorMaps)) {
      for (const p of oldOptions.colorMaps) {
        const color: string = p.color;
        if (color) {
          valuemap.options[p.text] = { color };
        }
      }
    }

    if (isArray(oldOptions.valueMaps)) {
      for (const p of oldOptions.valueMaps) {
        const text: string = p.text;
        const value: string = p.value;
        if (text && value) {
          let old = valuemap.options[value];
          if (old) {
            old.text = text;
          } else {
            valuemap.options[value] = { text };
          }
        }
      }
    }

    if (isArray(oldOptions.rangeMaps)) {
      for (const p of oldOptions.rangeMaps) {
        let from = +p.from;
        let to = +p.to;
        const text: string = p.text;
        if (text) {
          fieldConfig.defaults.mappings.push({
            type: MappingType.RangeToText,
            options: {
              from,
              to,
              result: { text },
            },
          });
        }
      }
    }

    // mutates the input
    panel.fieldConfig = fieldConfig;
  }

  return options;
};
