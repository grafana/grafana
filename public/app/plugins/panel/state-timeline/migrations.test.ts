import { PanelModel } from 'app/features/dashboard/state/PanelModel';

import { timelinePanelChangedHandler } from './migrations';

describe('Timeline Migrations', () => {
  it('from discrete panel', () => {
    const panel = {} as PanelModel;
    panel.options = timelinePanelChangedHandler(panel, 'natel-discrete-panel', { angular: discreteInV8 });
    expect(panel).toMatchSnapshot();
  });
});

const discreteInV8 = {
  id: 23763571993,
  gridPos: {
    h: 8,
    w: 12,
    x: 0,
    y: 0,
  },
  type: 'natel-discrete-panel',
  title: 'Panel Title',
  backgroundColor: 'rgba(128,128,128,0.1)',
  colorMaps: [
    {
      $$hashKey: 'object:365',
      color: '#7EB26D',
      text: '1',
    },
    {
      $$hashKey: 'object:366',
      color: '#EAB839',
      text: '20',
    },
    {
      $$hashKey: 'object:367',
      color: '#6ED0E0',
      text: '90',
    },
    {
      $$hashKey: 'object:368',
      color: '#EF843C',
      text: '30',
    },
    {
      $$hashKey: 'object:369',
      color: '#E24D42',
      text: '5',
    },
  ],
  crosshairColor: '#8F070C',
  display: 'timeline',
  extendLastValue: true,
  highlightOnMouseover: true,
  legendSortBy: '-ms',
  lineColor: 'rgba(0,0,0,0.1)',
  metricNameColor: '#000000',
  rangeMaps: [
    {
      $$hashKey: 'object:267',
      from: '1',
      text: 'AAA',
      to: '3',
    },
    {
      from: '4',
      to: '5',
      text: 'BBB',
      $$hashKey: 'object:544',
    },
  ],
  rowHeight: 50,
  showLegend: true,
  showLegendNames: true,
  showLegendPercent: true,
  showLegendValues: true,
  showTimeAxis: true,
  targets: [
    {
      refId: 'A',
      scenarioId: 'csv_metric_values',
      stringInput: '1,20,90,30,5,0',
    },
    {
      scenarioId: 'csv_metric_values',
      refId: 'B',
      stringInput: '1,20,30,5,0',
      hide: false,
    },
  ],
  textSize: 24,
  textSizeTime: 12,
  timeOptions: [
    {
      name: 'Years',
      value: 'years',
    },
    {
      name: 'Months',
      value: 'months',
    },
    {
      name: 'Weeks',
      value: 'weeks',
    },
    {
      name: 'Days',
      value: 'days',
    },
    {
      name: 'Hours',
      value: 'hours',
    },
    {
      name: 'Minutes',
      value: 'minutes',
    },
    {
      name: 'Seconds',
      value: 'seconds',
    },
    {
      name: 'Milliseconds',
      value: 'milliseconds',
    },
  ],
  timePrecision: {
    name: 'Minutes',
    value: 'minutes',
  },
  timeTextColor: '#d8d9da',
  units: 'currencyGBP',
  valueMaps: [
    {
      $$hashKey: 'object:265',
      op: '=',
      text: 'ONE',
      value: '111',
    },
    {
      value: '222',
      op: '=',
      text: 'TWO',
      $$hashKey: 'object:546',
    },
  ],
  valueTextColor: '#000000',
  writeLastValue: true,
  expandFromQueryS: 0,
  use12HourClock: false,
  useTimePrecision: false,
  writeAllValues: false,
  writeMetricNames: false,
  datasource: null,
};
