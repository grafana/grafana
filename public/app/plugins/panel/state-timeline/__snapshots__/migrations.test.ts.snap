// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Migrations from discrete panel 1`] = `
{
  "fieldConfig": {
    "defaults": {
      "custom": {
        "fillOpacity": 100,
        "lineWidth": 0,
      },
      "mappings": [
        {
          "options": {
            "1": {
              "color": "#7EB26D",
            },
            "111": {
              "text": "ONE",
            },
            "20": {
              "color": "#EAB839",
            },
            "222": {
              "text": "TWO",
            },
            "30": {
              "color": "#EF843C",
            },
            "5": {
              "color": "#E24D42",
            },
            "90": {
              "color": "#6ED0E0",
            },
          },
          "type": "value",
        },
        {
          "options": {
            "from": 1,
            "result": {
              "text": "AAA",
            },
            "to": 3,
          },
          "type": "range",
        },
        {
          "options": {
            "from": 4,
            "result": {
              "text": "BBB",
            },
            "to": 5,
          },
          "type": "range",
        },
      ],
      "unit": "currencyGBP",
    },
    "overrides": [],
  },
  "options": {
    "mergeValues": true,
  },
}
`;
